queasars/minimum_eigensolvers/evqe/evolutionary_algorithm/individual.py

```diff
@@ -0,0 +1,736 @@
+
+    def get_partially_parameterized_quantum_circuit(self, parameterized_layers: set[int]):
+        """Returns the quantum circuit as represented by this individual with only some
+        circuit layers being parameterized
+
+        :arg parameterized_layers: indices of the circuit layers which shall be parameterized
+        :type parameterized_layers: set[int]
+        :return: the partially parameterized QuantumCircuit
+        :rtype: QuantumCircuit
+        """
+        n_qubits: int = self.layers[0].n_qubits
+        circuit: QuantumCircuit = QuantumCircuit(n_qubits)
+        circuit.h(range(0, n_qubits))
+
+        # loop indices to allow negative numbers to refer to the last layers of the individual
+        parameterized_layers = {layer_id % len(self.layers) for layer_id in parameterized_layers}
+
+        # Apply each circuit layer one by one
+        for i, layer in enumerate(self.layers):
+            gate: Gate
+            # Get the layer as parameterized gate, if required
+            if i in parameterized_layers:
+                gate = layer.get_parameterized_layer_gate(layer_id=i)
+            else:
+                layer_parameter_values: tuple[float, ...] = tuple(
+                    parameter_value
+                    for j, parameter_value in enumerate(self.parameter_values)
+                    if j in self.layer_parameter_indices[i]
+                )
+                gate = layer.get_layer_gate(layer_id=i, parameter_values=layer_parameter_values)
+            circuit.append(instruction=gate, qargs=range(0, n_qubits))
+            circuit.barrier()
+        return circuit
+
+    def get_parameter_values(self) -> tuple[float, ...]:
+        return self.parameter_values
+
+    def get_layer_parameter_values(self, layer_id: int) -> tuple[float, ...]:
+        """
+        Returns the parameter values for a specified circuit layer
+
+        :param layer_id: index of the layer
+        :type layer_id: int
+        :return: the parameter values
+        :rtype: tuple[float, ...]
+        """
+        # loop indices to allow negative numbers to refer to the last layers of the individual
+        layer_id = layer_id % len(self.layers)
+
+        layer_parameter_values: tuple[float, ...] = tuple(
+            parameter_value
+            for i, parameter_value in enumerate(self.parameter_values)
+            if i in self.layer_parameter_indices[layer_id]
+        )
+        return layer_parameter_values
+
+    def get_n_controlled_gates(self) -> int:
+        """Get the amount of controlled gates over all circuit layers of this individual
+
+        :return: the number of controlled gates
+        :rtype: int
+        """
+        return sum(layer.n_controlled_gates for layer in self.layers)
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
+
+    def __hash__(self):
+        return hash((self.n_qubits, self.layers, self.parameter_values))
```

Comment:
Aren't `__hash__` and `__eq__` already implemented as you defined this as a dataclass with `Frozen=True`. In particular the implementation will be pretty much identical (at least for the hash function).

Reply:
Ah right, I found the discussion there. Interesting that the auto implemented methods seem to have a different signature (or at least doesn't implement the abstract method), that's surprising to me.
Conceptually it seems appropriate to me that `EVQEIndividual` is a dataclass, though 🤔 .
Also from the documentation it is ok to implement these methods yourself (the decorators just help you to avoid the need to do this yourself), so from my side you can keep it this way.